### PR TITLE
[Flax] Add remat (gradient checkpointing)

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Adds gradient checkpointing in Flax (_c.f._ #17399). The API currently takes the form of a method:

```python
from transformers import BertConfig, FlaxBertModel

model = FlaxBertModel(BertConfig())
model.enable_gradient_checkpointing()
```

Note: checkpointing has currently only been implemented for FlaxBert. Implementing for all Flax models is a TODO.

TODO:

- [x] Add checkpointing to `init`
- [x] Add checkpointing to `from_pretrained`
- [x] Add model tests for FlaxBert in `test_modeling_flax_bert`
- [x] Decide on API: checkpointing with a kwarg (`gradient_checkpointing=True`) or a method (`model.gradient_checkpointing_enable()`)?
- [x] Add API functionality for remat policies (c.f. https://github.com/google/jax/blob/636345fd67758c19c5345bee2301df34b6f1c540/jax/_src/ad_checkpoint.py#L44)
- [ ] Copy checkpointing logic to all Flax models
- [ ] Move model tests to `test_modeling_flax_common`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc @borisdayma

### CODE REVIEW

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 1]
```diff
@@ -582,12 +590,12 @@ def __call__(
...
-                encoder_attention_mask=encoder_attention_mask,
-                init_cache=init_cache,
-                deterministic=deterministic,
-                output_attentions=output_attentions,
+                head_mask[i] if head_mask is not None else None,
+                encoder_hidden_states,
+                encoder_attention_mask,
+                init_cache,
+                deterministic,
+                output_attentions,
```
Comment (sanchit-gandhi) [inline src/transformers/models/bert/modeling_flax_bert.py:598]: Note: `remat` does not support kwargs, hence the need to change to args

**Comment by sanchit-gandhi** [comment]
> Is there an inconvenient in adding it to all layers?

We're doing this! By wrapping `FlaxBertLayer` in a `remat` operation, each Bert layer (attention, intermediate FF, final FF + optional cross-attention layers) has `remat` applied to it:
https://github.com/huggingface/transformers/blob/ea8150a5f932ab28efbe2e7a31fee1ca77c289a5/src/transformers/models/bert/modeling_flax_bert.py#L555
We then use this remat'd layer to construct the Transformer block (layers collection):
https://github.com/huggingface/transformers/blob/ea8150a5f932ab28efbe2e7a31fee1ca77c289a5/src/transformers/models/bert/modeling_flax_bert.py#L559-L562
Meaning that each component of the Bert layer is checkpointed, and that **all** Bert layers in the Transformer block (layers collection) are checkpointed.

File: src/transformers/models/bert/modeling_flax_bert.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py:601]: ok!

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 2]
```diff
@@ -617,9 +628,16 @@ def __call__(
 class FlaxBertEncoder(nn.Module):
     config: BertConfig
     dtype: jnp.dtype = jnp.float32  # the dtype of the computation
+    gradient_checkpointing: bool = False
+    remat_policy: Callable[..., bool] = (None,)  # the gradient checkpointing policy
```
Comment (patrickvonplaten) [inline src/transformers/models/bert/modeling_flax_bert.py:632]: Are there multiple policies? Would one every use another one then the default one? Wondering if allowing this parameter to be customizable might be a bit scary for the user and make the whole functionality less understandable. Think I'd prefer to just use the default here and not allow the user to configure it

**Review by patrickvonplaten — APPROVED** [review]
Very nice! Also cc @younesbelkada 

We could also look into implementing this for OPT and BLOOM in Flax :-) Great job @sanchit-gandhi 

Only feedback from my side would be to remove the option to overwrite the policy (also since we don't test it)

File: src/transformers/models/bert/modeling_flax_bert.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/bert/modeling_flax_bert.py]: Thanks for the lengthy explanation! If different policies can work best depending on the circumstances, I'm happy to leave as is :-)

**Comment by patrickvonplaten** [comment]
Cool! Once the tests are green, happy to merge it here :-)

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 3]
```diff
@@ -544,10 +547,18 @@ def __call__(
 class FlaxBertLayerCollection(nn.Module):
     config: BertConfig
     dtype: jnp.dtype = jnp.float32  # the dtype of the computation
+    gradient_checkpointing: bool = False
+    remat_policy: Callable[..., bool] = (None,)  # the gradient checkpointing policy
 
     def setup(self):
+        if self.gradient_checkpointing:
+            FlaxBertCheckpointLayer = remat(FlaxBertLayer, static_argnums=(5, 6, 7), policy=self.remat_policy)
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/bert/modeling_flax_bert.py]: `remat` prevented re-use of the class name `FlaxBertLayer`:
https://github.com/google/flax/discussions/2251
Can re-name in a follow-up PR if we find a workaround!
